Adding Python3 Compatibility

bushbrother opened the pull request.

bushbrother committed:
Updated to support Python3, Added Python EoL notification

bushbrother committed:
Update README.md

hugbug committed:
removed unnecessary changes

hugbug commented:
Thank you.
I've removed the EOL stuff and the unnecessary formatting changes (tabs vs spaces). Now we can concentrate on the relevant changes.

Are you sure HTTP requests are handled properly? All references to `urllib2` remained in the code but this module isn't imported as a whole on python2 and not at all on python3.

bushbrother commented:
OK agree on EOL stuff :)

urllib is handled here, it tries Python3 way, if that errors then it imports Python2 way - this is the recommended way at python.org if you are unable to use the "future" library that requires a pip install:

Python2 urllib handling
try:
    from urllib.parse import urlparse, urlencode
    from urllib.request import urlopen, Request
    from urllib.error import HTTPError
except ImportError:
    from urlparse import urlparse
    from urllib import urlencode
    from urllib2 import urlopen, Request, HTTPError

hugbug commented:
That's only imports but the further code in the script refers to `urllib2` which isn't available on python 3. Are you sure the script in its current state works on python 3? I think it doesn't but I haven't tested.

bushbrother commented:
Gotcha :) - Let me verify.

bushbrother committed:
Update FailureLink.py

hugbug commented:
Shouldn't it be enough to replace `urllib2.urlopen` with `urlopen` (and other cases similarly)?

bushbrother committed:
Update FailureLink.py

bushbrother committed:
Update FailureLink.py

bushbrother committed:
Update FailureLink.py

bushbrother committed:
Update FailureLink.py

bushbrother committed:
Update FailureLink.py

bushbrother committed:
Update FailureLink.py

bushbrother committed:
Update FailureLink.py

bushbrother committed:
Update FailureLink.py

bushbrother commented:
You were quite right ... sorry I am very new to Python :) Takes me ages to get used to! Should be ok now ...

bushbrother committed:
Switch to getenv to avoid errors if value not present

bushbrother committed:
Update FailureLink.py

bushbrother committed:
Restoring

bushbrother committed:
Update FailureLink.py

bushbrother committed:
Update FailureLink.py

bushbrother committed:
Fix os.environ errors (http://www.snowcrash.eu/python-raise-keyerrorkey-from-none/)

bushbrother commented:
Apologies again, while that fixed the http errors somehow I had missed environmental issues. Please use this latest commit.

bushbrother commented:
Hi - Any word on this? I have been using for a while now and seems OK.

jpartain89 commented:
Is there anyway someone can "Ok" this? Docker containers have started to mess up due to everyone moving to Python3 and this is the main reason for our containers not working now.

cxhamilton commented on FailureLink.py:
```diff
@@ -56,102 +56,121 @@
 # NZB-file of the same release and add it to queue. If disabled the indexer
 # site is still informed about the failure but no other nzb-file is queued.
-#DownloadAnotherRelease=no
+# DownloadAnotherRelease=no
```
I'm running 21.0 stable and the space between the "#" and "DownloadAnotherRelease=no" breaks the configuration screen for this extension script. The same comment applies to all of the config properties.

bushbrother replied:
Hmmm you are right. Will fix.

bushbrother committed:
Update FailureLink.py

Fix spaces causing config options not to appear

cxhamilton commented:
@hugbug Thoughts on this?

jpartain89 commented:
Anyone?

hugbug commented:
Hi guys.
This is a one file script. If there is a python3 version of script - please just use it, you don't have to wait for me to merge the PR.
As for the PR - it's difficult to check what was changed because the file was reformatted. Please restore the original formatting. Thanks.

bushbrother committed:
Update FailureLink.py

Attempt to highlight only changed lines ready for pull.

bushbrother commented:
@hugbug - I pushed an update so now it only shows the edited lines ... not sure how that happened :) - Should now show the changes to all the "print" lines, the library handling section and the parts of the code that needed to be tweaked to use new libraries.

bushbrother committed:
Update FailureLink.py

Reverting back also reintroduced issues with tab/line inconsistencies which only Python3 compains about.

bushbrother commented:
Also then had to fix spaces/tab inconsistencies that re-introduced ... Python3 is picky about them!

hugbug commented:
Thank you.
Before I merge this - can you assure me that the script is working and was tested in both Python 3 and 2?

bushbrother commented:
> Thank you.
> Before I merge this - can you assure me that the script is working and was tested in both Python 3 and 2?

To be honest - I am confident it works on Py3.x as that is what I am using - but not 100% sure on Py2.x. Let me go away and spin up a container using older Python and confirm :)

bushbrother commented:
OK - I created 2 containers running NZBGet, one built with Py3.x and another with Py2.7.x - I cloned my fork (with these Py3 changes) to both script directories and tried a known bad NZB on both - they both run correctly and trigger a new file to the indexer (dognzb) - however the indexer is giving a HTTP500 error. I checked the link generated and put it in my browser - it also suffers the same thing. So now I am looking to see if you know of other indexers supporting this header so I can try them? Dognzb seems to be broken ...

The script completes fine for a good download, and appears to be working otherwise ...

cxhamilton commented:
I can confirm the 500. I raised the issue on the DOGnzb discussion forum a couple of weeks ago. No response so far.

hugbug commented:
Thanks for testing. I'm merging the PR.
I do not know other indexers supporting this too :(
